drgn/helpers/linux/cgroup.py

```diff
@@ -0,0 +1,183 @@
+# SPDX-License-Identifier: GPL-3.0+
+
+"""
+Cgroup
+------
+
+The ``drgn.helpers.linux.cgroup`` module provides helpers for working with
+cgroup interface in :linux:`include/linux/cgroup.h`. Only cgroup v2 is
+supported.
+"""
+
+
+import enum
+
+from drgn import cast, container_of
+from drgn.helpers.linux.kernfs import kernfs_name, kernfs_path
+from drgn.helpers.linux.list import list_for_each_entry
+
+
+__all__ = [
+    'cgroup_name',
+    'cgroup_parent',
+    'cgroup_path',
+    'css_for_each_child',
+    'css_for_each_descendant_pre',
+    'css_next_child',
+    'css_next_descendant_pre',
+    'sock_cgroup_ptr',
+]
+
+
+def sock_cgroup_ptr(skcd):
+    """
+    .. c:function:: struct cgroup *sock_cgroup_ptr(struct sock_cgroup_data *skcd)
+
+    Get cgroup of a socket by ``struct sock_cgroup_data`` member embedded to
+    every ``struct sock``.
+
+
+    :return: ``struct cgroup *`` object.
+    """
+    return cast('struct cgroup *', skcd.val)
+
+
+def cgroup_parent(cgrp):
+    """
+    .. c:function:: struct cgroup *cgroup_parent(struct cgroup *cgrp)
+
+    Get parent of cgroup if it exists.
+
+    :return: ``struct cgroup`` object or None.
+    """
+    parent_css = cgrp.self.parent
+    if parent_css:
+        return container_of(parent_css, 'struct cgroup', 'self')
+    return None
+
+
+def cgroup_name(cgrp):
+    """
+    .. c:function:: char *cgroup_name(struct cgroup *cgrp)
+
+    Get cgroup name.
+
+    :rtype: bytes
+    """
+    return kernfs_name(cgrp.kn)
+
+
+def cgroup_path(cgrp):
+    """
+    .. c:function:: char *cgroup_path(struct cgroup *cgrp)
+
+    Get cgroup path.
+
+    :rtype: bytes
+    """
+    return kernfs_path(cgrp.kn)
+
+
+def _get_css_flags(prog):
```

Comment:
Right, it's only for internal usage.

The main reason I did it is to cache the enum to avoid `read()`ing it from kern mem all the time during iteration, but I admin that I don't know the whole picture when drgn reads and when it doesn't so maybe it's premature optimization.

I'll remove it but lmk if caching makes sense here.

Reply:
Enums are available in the kernel debug info, so we don't have to read kernel memory for it. So in this case, caching won't help much. I should document some of the low-level details of drgn that can help people writing helpers :)